src/token/erc1155/erc1155.cairo

```diff
@@ -0,0 +1,545 @@
+        /// Will mint (or burn) if `from` (or `to`) is the zero address.
+        ///
+        /// Requirements:
+        ///
+        /// - `to` is either an account contract or supports the `IERC1155Receiver` interface.
+        /// - `token_ids` and `values` must have the same length.
+        ///
+        /// Emits a `TransferSingle` event if the arrays contain one element, and `TransferBatch` otherwise.
+        ///
+        /// NOTE: The ERC-1155 acceptance check is not performed in this function.
+        /// See `update_with_acceptance_check` instead.
+        fn update(
+            ref self: ComponentState<TContractState>,
+            from: ContractAddress,
+            to: ContractAddress,
+            token_ids: Span<u256>,
+            values: Span<u256>
+        ) {
+            assert(token_ids.len() == values.len(), Errors::INVALID_ARRAY_LEN);
+
+            let mut index = 0;
+            loop {
+                if index == token_ids.len() {
+                    break;
+                }
+                let token_id = *token_ids.at(index);
+                let value = *values.at(index);
+                if from.is_non_zero() {
+                    let from_balance = self.ERC1155_balances.read((token_id, from));
+                    assert(from_balance >= value, Errors::INSUFFICIENT_BALANCE);
+                    self.ERC1155_balances.write((token_id, from), from_balance - value);
+                }
+                if to.is_non_zero() {
+                    let current = self.ERC1155_balances.read((token_id, to));
+                    self.ERC1155_balances.write((token_id, to), current + value);
+                }
+                index += 1;
+            };
+
+            let operator = get_caller_address();
+            if token_ids.len() == 1 {
+                self
+                    .emit(
+                        TransferSingle {
+                            operator, from, to, id: *token_ids.at(0), value: *values.at(0)
+                        }
+                    );
+            } else {
+                self.emit(TransferBatch { operator, from, to, ids: token_ids, values });
+            }
+        }
+
+        /// Version of `update` that performs the token acceptance check by calling
+        /// `IERC1155Receiver-onERC1155Received` or `IERC1155Receiver-onERC1155BatchReceived` if
+        /// the receiver is not reconized as an account.
+        fn update_with_acceptance_check(
+            ref self: ComponentState<TContractState>,
+            from: ContractAddress,
+            to: ContractAddress,
+            token_ids: Span<u256>,
+            values: Span<u256>,
+            data: Span<felt252>
+        ) {
+            self.update(from, to, token_ids, values);
+            if token_ids.len() == 1 {
+                _check_on_ERC1155_received(from, to, *token_ids.at(0), *values.at(0), data);
+            } else {
+                _check_on_ERC1155_batch_received(from, to, token_ids, values, data);
```

Comment:
should we add custom errors?

Reply:
i think we should add them, for consistency with ERC721 and [our Solidity module](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC1155/utils/ERC1155Utils.sol#L34-L39)

Reply:
We do have them already. I changed the message to safe transfer failed for consistency.

Reply:
In ERC721 we are returning a bool from the check method, while the doc comment states the transaction will revert. In this module I added the assertion inside the check function because of this.

Reply:
we should amend the docs to reflect the current implementation. in fact that's the in-code docs, we need to include the method in the docsite

i think returning bool is the right call for the `_check` functions (otherwise i'd call them `assert_`), the problem resides in `update_with_acceptance_check` not asserting the returned bool, not even bubbling it up so it can be used by the calling method

Reply:
my bad! i was looking at an older commit. I see, the check is in place just that it's within the `_check` functions, what do you think about mimicking 721 and move the checks to the `_with_acceptance_check` functions?

Reply:
No issue with that, but as mentioned in my previous comment, I think we should update the doc comment, because is a bit missleading.